Examples/Io/Csv/src/CsvBFieldWriter.cpp

```diff
@@ -0,0 +1,261 @@
+          "Input B-field does not support 3-dimensional indexing.");
+    }
+
+    /*
+     * If our magnetic field is grid-like, we know that it has a built-in size
+     * and bin count. We can use these, but the user may want to override the
+     * values with custom values.
+     */
+    for (std::size_t i = 0; i < ConfigType::NDims; ++i) {
+      /*
+       * For each dimension, we first get the corresponding value from the
+       * grid-based vector field...
+       */
+      bins[i] = field.getNBins()[i];
+      min[i] = field.getMin()[i];
+      max[i] = field.getMax()[i];
+
+      /*
+       * ...and then we override them with the optional user-supplied values.
+       */
+      if (config.bins[i]) {
+        bins[i] = *config.bins[i];
+      }
+
+      if (config.range[i][0]) {
+        min[i] = *config.range[i][0];
+      }
+
+      if (config.range[i][1]) {
+        max[i] = *config.range[i][1];
+      }
+    }
+  } else {
+    /*
+     * If the vector field is not grid based, then there is no side and we are
+     * forced to use the user-supplied data. Remember that in this case, the
+     * values are not optional.
+     */
+    for (std::size_t i = 0; i < ConfigType::NDims; ++i) {
+      bins[i] = config.bins[i];
+      min[i] = config.range[i][0];
+      max[i] = config.range[i][1];
+    }
+  }
+
+  /*
+   * Next, we calculate the size (in physical space) of each bin.
+   */
+  Vector delta;
+
+  for (std::size_t i = 0; i < ConfigType::NDims; ++i) {
+    delta[i] = (max[i] - min[i]) / bins[i];
+  }
+
+  /*
+   * Create the appropriate magnetic field context and cache to interact with
+   * the B fields.
+   */
+  Acts::MagneticFieldContext mctx{};
+  typename FieldType::Cache cache = field.makeCache(mctx);
+
+  /*
+   * Finally, we can begin to fill the output file with data from our B field.
+   * Again, the procedure is slightly different depending on whether we are
+   * working with Cartesian or cylindrical coordinates.
+   */
+  if constexpr (Coord == CoordinateType::XYZ) {
+    ACTS_INFO("Writing XYZ field of size "
+              << (bins[0] + 1) << " x " << (bins[1] + 1) << " x "
+              << (bins[2] + 1) << " to file " << config.fileName << "...");
+
+    std::size_t total_items = (bins[0] + 1) * (bins[1] + 1) * (bins[2] + 1);
+
+    /*
+     * For Cartesian coordinates, iterate over bins in the x, y, and z
+     * directions. Note that we iterate one additional time because we are
+     * writing the _edges_ of the bins, and the final bin needs to be closed.
+     */
+    for (std::size_t x = 0; x <= bins[0]; ++x) {
+      for (std::size_t y = 0; y <= bins[1]; ++y) {
+        for (std::size_t z = 0; z <= bins[2]; ++z) {
+          /*
+           * Compute the geometric position of this bin, then request the
+           * magnetic field vector at that position.
+           */
+          Acts::Vector3 pos = {x * delta[0] + min[0], y * delta[1] + min[1],
+                               z * delta[2] + min[2]};
```

Comment:
How happy are you with the performance? Would it make sense to rewrite this part to something like this?
```
Acts::Vector3 pos = { min_x-delta_x, min_y-delta_y, min_z-delta_z }
for (x-bins)
  pos = { +delta_x, ..., ... }
  for (y-bins)
    pos = {..., +delta_y, ... }
    for (z-bins)
      pos = { ..., ..., +delta_z }
```

(same for cylindrical)

Reply:
Hmm, this is an interesting suggestion, but my issue with this is floating point error. If we go for a loop structure like this, if we iterate 300 times, then we incur floating point precision loss 300 times instead of just once, which is probably not ideal. I don't know how big the impact of this would be in practice, but this tool is really a "use once" sort of tool, because you don't need to dump B-fields to disk every day!

Reply:
I believe sequential adds are less accurate than multiplying. `pos` could drift away because of stacked FP inaccuracies